Hook up IPC for webNavigation APIs

b-weinstein opened the pull request:
#### 0575d062092c09d95db62b0064342021dc06366f
<pre>
Hook up IPC for webNavigation APIs
<a href="https://bugs.webkit.org/show_bug.cgi?id=264595">https://bugs.webkit.org/show_bug.cgi?id=264595</a>
<a href="https://rdar.apple.com/118237023">rdar://118237023</a>

Reviewed by Timothy Hatcher.

This patch creates a new WebExtensionFrameParameters and adopts it to pass frame data from the UI process
back to the Web process.

* Source/WebKit/DerivedSources-input.xcfilelist: Add the new file.
* Source/WebKit/DerivedSources.make: Ditto.
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared): Ditto.
* Source/WebKit/Shared/Extensions/WebExtensionFrameParameters.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionFrameParameters.serialization.in: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWebNavigationCocoa.mm: Added.
(WebKit::WebExtensionContext::webNavigationGetFrame): Call into the UI process and generate Objective-C objects based on the return value.
(WebKit::WebExtensionContext::webNavigationGetAllFrames): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h: Add the message handling methods.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in: Ditto.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Add the new files.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationCocoa.mm:
(WebKit::toWebAPI): Turn the passed in arguments into NS types.
(WebKit::WebExtensionAPIWebNavigation::getAllFrames): Call into the UI process to perform the work.
(WebKit::WebExtensionAPIWebNavigation::getFrame): Ditto.

Canonical link: <a href="https://commits.webkit.org/270564@main">https://commits.webkit.org/270564@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e469639c324c48695e9d6103330ee2fed693ec23

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25773 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4380 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27057 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27871 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/23602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6139 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1815 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23714 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26023 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3288 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/22211 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28451 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/25937 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/2910 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23522 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/23538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/27100 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2927 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1161 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4315 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6198 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3380 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3243 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWebNavigationCocoa.mm:
```diff
@@ -0,0 +1,80 @@
+        return;
+    }
+
+    // FIXME: Implement this.
```
radar number?

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWebNavigationCocoa.mm:
```diff
@@ -0,0 +1,80 @@
+{
+    auto tab = getTab(tabIdentifier);
+    if (!tab) {
+        completionHandler({ });
```
should the `!tab` and `webView` cases set `errorOccurred` for the `WebExtensionFrameParameters`?

b-weinstein replied:
I don't think so. From Chrome's documentation:

errorOccurred
boolean
True if the last navigation in this frame was interrupted by an error, i.e. the onErrorOccurred event fired.

In this case, not being able to find a tab or page isn't an error. According to the docs we should be returning nil:
![Screenshot 2023-11-10 at 9 11 59 AM](https://github.com/WebKit/WebKit/assets/114437188/d6203d57-e591-444a-859f-770f8dfd4642)

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationCocoa.mm:
```diff
@@ -28,25 +28,100 @@
+
+    result[errorOccurredKey] = frameInfo.errorOccurred ? @(frameInfo.errorOccurred.value()) : @NO;
+    result[parentFrameIdKey] = @(toWebAPI(frameInfo.parentFrameIdentifier));
+    result[urlKey] = frameInfo.url ? (NSString *)frameInfo.url.value().string() : @"";
```
i think alternatively this could work?
`result[urlKey] = (NSString *)frameInfo.url.value_or(""_s)`

xeenon replied:
Regarding @kiaraarose's comment about using `value_or()`, in this case it would not work, since `value()` is `URL`.

Tabs does this, to account for null URLs so they are returned as empty strings too.

```
if (parameters.url)
    result[urlKey] = !parameters.url.value().isNull() ? (NSString *)parameters.url.value().string() : emptyURLValue;
```

The `url` key is optional in tabs, so you might need to format differently to handle the optional and isNull on URL if `url` is a required key for frame info.

Like this:

```
result[urlKey] = parameters.url && !parameters.url.value().isNull() ? (NSString *)parameters.url.value().string() : emptyURLValue;
```

b-weinstein replied:
I think the current structure is a bit more readable, but don't feel strongly about it.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionFrameParameters.h:
```diff
@@ -0,0 +1,45 @@
+namespace WebKit {
+
+struct WebExtensionFrameParameters {
+
```
Drop the empty line.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionFrameParameters.h:
```diff
@@ -0,0 +1,45 @@
+    std::optional<URL> url;
+    WebKit::WebExtensionFrameIdentifier parentFrameIdentifier;
+    std::optional<WebKit::WebExtensionFrameIdentifier> frameIdentifier;
+
```
Drop the empty line.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionFrameParameters.serialization.in:
```diff
@@ -0,0 +1,34 @@
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+struct WebKit::WebExtensionFrameParameters {
+
```
Ditto.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionFrameParameters.serialization.in:
```diff
@@ -0,0 +1,34 @@
+    std::optional<URL> url;
+    WebKit::WebExtensionFrameIdentifier parentFrameIdentifier;
+    std::optional<WebKit::WebExtensionFrameIdentifier> frameIdentifier;
+
```
Ditto.

kiaraarose commented on Source/WebKit/Shared/Extensions/WebExtensionFrameParameters.serialization.in:
```diff
@@ -0,0 +1,34 @@
+
+struct WebKit::WebExtensionFrameParameters {
+
+    std::optional<bool> errorOccurred;
```
could make this not optional and default it to `false` so you only have to do `frameInfo.errorOccurred` in the `toWebAPI()`

xeenon replied:
Yeah, `bool errorOccurred { false };`

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationCocoa.mm:
```diff
@@ -28,25 +28,100 @@
+        return;
+
+    NSNumber *tabId = details[tabIdKey];
+    std::optional<WebExtensionTabIdentifier> tabIdentifier = toWebExtensionTabIdentifier(tabId.doubleValue);
```
`auto tabIdentifier = toWebExtensionTabIdentifier(tabId.doubleValue);`

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationCocoa.mm:
```diff
@@ -28,25 +28,100 @@
+        return;
+
+    NSNumber *frameId = details[frameIdKey];
+    std::optional<WebExtensionFrameIdentifier> frameIdentifier = toWebExtensionFrameIdentifier(frameId.doubleValue);
```
`auto frameIdentifier = toWebExtensionFrameIdentifier(frameId.doubleValue);`

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationCocoa.mm:
```diff
@@ -28,25 +28,100 @@
+        return;
+
+    NSNumber *tabId = details[tabIdKey];
+    std::optional<WebExtensionTabIdentifier> tabIdentifier = toWebExtensionTabIdentifier(tabId.doubleValue);
```
ditto on using `auto`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationCocoa.mm:
```diff
@@ -28,25 +28,100 @@
+    for (auto frame : allFrames)
+        [result addObject:toWebAPI(frame)];
+
+    return [result copy];
```
```
    return createNSArray(allFrames, [](auto& frame) {
        return toWebAPI(frame);
    }).get();
```

xeenon replied:
Might need `#import <wtf/cocoa/VectorCocoa.h>` for this.

b-weinstein replied:
Neat!

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWebNavigationCocoa.mm:
```diff
@@ -0,0 +1,80 @@
+{
+    auto tab = getTab(tabIdentifier);
+    if (!tab) {
+        completionHandler({ });
```
`std::nullopt` for the error cases would be good, to be clear what this is.

b-weinstein committed:
Hook up IPC for webNavigation APIs
https://bugs.webkit.org/show_bug.cgi?id=264595
rdar://118237023

Reviewed by Timothy Hatcher.

This patch creates a new WebExtensionFrameParameters and adopts it to pass frame data from the UI process
back to the Web process.

* Source/WebKit/DerivedSources-input.xcfilelist: Add the new file.
* Source/WebKit/DerivedSources.make: Ditto.
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared): Ditto.
* Source/WebKit/Shared/Extensions/WebExtensionFrameParameters.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionFrameParameters.serialization.in: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWebNavigationCocoa.mm: Added.
(WebKit::WebExtensionContext::webNavigationGetFrame): Call into the UI process and generate Objective-C objects based on the return value.
(WebKit::WebExtensionContext::webNavigationGetAllFrames): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h: Add the message handling methods.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in: Ditto.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Add the new files.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationCocoa.mm:
(WebKit::toWebAPI): Turn the passed in arguments into NS types.
(WebKit::WebExtensionAPIWebNavigation::getAllFrames): Call into the UI process to perform the work.
(WebKit::WebExtensionAPIWebNavigation::getFrame): Ditto.

Canonical link: https://commits.webkit.org/270564@main
